### PR TITLE
use separate manifest files for delete files

### PR DESCRIPTION
**Pull request by JanKaul**
This PR implements an alternative approach to solve #212.

It selects two manifest files, one for data and one for delete files. The aim is to limit the number of additional manifest files that have to be created.

### CODE REVIEW

File: datafusion_iceberg/tests/equality_delete.rs [hunk 1]
```diff
@@ -194,4 +203,28 @@ pub async fn test_equality_delete() {
...
+        .object_store()
+        .get(&object_store::path::Path::from(format!(
+            "{table_dir}/metadata/version-hint.text"
+        )))
+        .await
+        .unwrap()
+        .bytes()
+        .await
+        .unwrap();
+    let latest_version_str = std::str::from_utf8(&latest_version).unwrap();
```
Comment (gruuya) [inline datafusion_iceberg/tests/equality_delete.rs:217]: This was a bit unfortunate, since i realized that:
- just loading the table won't give you the full path to latest metadata file
- the `version-hint.text` (which does contain the full path) itself  is not readable by duckdb
   ```
   Iceberg metadata file not found for table version '/Users/gruuya/EDB/store/warehouse/tpch/query12/. metadata/00000-3f569e94-5601-48f3-9199-8d71df4ea7b0.metadata.json' using 'none' compression and format(s): 'v%s%s.metadata.json,%s%s.metadata.json'
   ```

There are ways to [customize](https://duckdb.org/docs/stable/core_extensions/iceberg/overview.html#working-with-alternative-metadata-naming-conventions) how duckdb parses the version to read, but i think those work only when the version file is not detected in the first place. Either way it's a separate issue.

File: iceberg-rust/src/table/transaction/append.rs [hunk 2]
```diff
@@ -127,70 +129,122 @@ pub(crate) fn select_manifest_partitioned(
...
+                    continue;
+                };
+
+                if selected_row_count.is_some_and(|x| x > row_count) {
+                    manifest_list_writer.append_ser(selected_manifest)?;
+                    selected_delete_state = Some((Some(row_count), manifest));
+                    continue;
+                } else {
+                    manifest_list_writer.append_ser(manifest)?;
+                    continue;
```
Comment (gruuya) [inline iceberg-rust/src/table/transaction/append.rs:235]: These `continue`s are redundant; likewise above (lines 143, 147, 161, 165, 211, 214).

Besides that I find the let-some-else-continue less readable so i refactored those in my pr, though that's not a big deal.

File: iceberg-rust/src/table/manifest_list.rs [hunk 3]
```diff
@@ -664,14 +671,19 @@ impl<'schema, 'metadata> ManifestListWriter<'schema, 'metadata> {
     ///     object_store,
     /// ).await?;
     /// ```
-    pub(crate) async fn append_filtered_and_finish(
-        mut self,
+    pub(crate) async fn append_filtered(
```
Comment (gruuya) [inline iceberg-rust/src/table/manifest_list.rs:681]: All these need docs updates that it doesn't `finish` on it's own now.

File: iceberg-rust/src/table/manifest_list.rs [hunk 4]
```diff
@@ -395,7 +398,8 @@ impl<'schema, 'metadata> ManifestListWriter<'schema, 'metadata> {
         Ok(Self {
             table_metadata,
             writer,
-            selected_manifest: Some(manifest),
+            selected_data_manifest: Some(data_manifest),
+            selected_delete_manifest: delete_manifest,
```
Comment (gruuya) [inline iceberg-rust/src/table/manifest_list.rs:404]: I like that there are both data and delete manifest entries selected. It's more explicit and better than in my pr where i only pick to continue in the one which has more new files to be added to.

File: iceberg-rust/src/table/transaction/append.rs [hunk 5]
```diff
@@ -127,70 +129,122 @@ pub(crate) fn select_manifest_partitioned(
...
         }
     }
-    selected_state
-        .map(|(_, entry)| SelectedManifest {
-            manifest: entry,
-            file_count_all_entries,
-        })
-        .ok_or(Error::NotFound("Manifest for insert".to_owned()))
+    let (_, data_manifest) =
+        selected_data_state.ok_or(Error::NotFound("Manifest for insert".to_owned()))?;
```
Comment (gruuya) [inline iceberg-rust/src/table/transaction/append.rs:241]: Maybe this shouldn't be an error but instead, the entire `Option<ManifestListEntry>` should be returned, since the caller (`ManifestListWriter::from_existing`) knows how to handle that correctly.

Granted I think this scenario can only happen if we have an empty/new table and we wont to append some deletes to it, or some such edge case.

**Comment by gruuya** [comment]
Btw those two failing mat-view tests are solved by this https://github.com/JanKaul/iceberg-rust/pull/227/commits/0dc313a2e273cf155cafc7b0940be846a6c5c7ea (otherwise you end up generating identical data and delete manifest paths).

File: datafusion_iceberg/tests/equality_delete.rs (hunk 1 above)
Comment (JanKaul) [reply datafusion_iceberg/tests/equality_delete.rs:217]: Hmm, is the naming of the metadata.json wrong then?

File: iceberg-rust/src/table/transaction/append.rs (hunk 2 above)
Comment (JanKaul) [reply iceberg-rust/src/table/transaction/append.rs:235]: I'll have look, thanks. I wanted to quickly sketch out the general idea. Haven't looked into that yet.

**Comment by JanKaul** [comment]
> Btw those two failing mat-view tests are solved by this [0dc313a](https://github.com/JanKaul/iceberg-rust/commit/0dc313a2e273cf155cafc7b0940be846a6c5c7ea) (otherwise you end up generating identical data and delete manifest paths).

Awesome, thanks! I was stuck on how to solve that one.

File: datafusion_iceberg/tests/equality_delete.rs (hunk 1 above)
Comment (gruuya) [reply datafusion_iceberg/tests/equality_delete.rs:217]: I think the naming is fine, it's just that version-hint.text should contain `00000-3f569e94-5601-48f3-9199-8d71df4ea7b0` (e.g. for a file named `00000-3f569e94-5601-48f3-9199-8d71df4ea7b0.metadata.json`), so that it complies with the format `%s%s.metadata.json` (the other %s is for the compression codec.

I also know spark doesn't expect that format either
```python
In [30]: df = spark.read.format("iceberg").load("/Users/gruuya/EDB/store/test/orders")
25/07/16 09:12:42 WARN HadoopTableOperations: Error reading version hint file /Users/gruuya/EDB/store/test/orders/metadata/version-hint.text
java.lang.NumberFormatException: For input string: "/Users/gruuya/EDB/store/test/orders/metadata/v2.metadata.json"
```
but I'm not sure exactly what's the proper format for it then.

File: iceberg-rust/src/table/transaction/operation.rs [hunk 6]
```diff
@@ -138,40 +138,82 @@ impl Operation {
...
-                    manifest_list_writer
-                        .append_multiple_and_finish(
-                            new_datafile_iter,
-                            snapshot_id,
-                            n_splits,
-                            object_store,
-                        )
-                        .await?
-                };
+                if n_data_files != 0 {
```
Comment (gruuya) [inline iceberg-rust/src/table/transaction/operation.rs]: nit: could use something like  
```rust
for (content, files) in [(Content::Data, data_files), (Content::Deletes, delete_files)]
```
to deduplicate some code below.

**Review by gruuya — APPROVED** [review]

**Comment by gruuya** [comment]
Fwiw, I've also been seeing the [no-disk-space](https://github.com/JanKaul/iceberg-rust/actions/runs/16770800255/job/47485232178?pr=228#step:6:207) error during trino image pull.

I've done some diagnosing on it and it seems like `/usr/local` is inflated, and just by removing `/usr/local/lib/android` around 10GB could be [reclaimed](https://github.com/JanKaul/iceberg-rust/actions/runs/16745871619/job/47404175959#step:6:56).

File: iceberg-rust/src/table/transaction/append.rs (hunk 5 above)
Comment (JanKaul) [reply iceberg-rust/src/table/transaction/append.rs:241]: I think think there is no valid scenario where there shouldn't be a data manifest selected here. And I think not having an `Option` in `SelectedManifest` is the most explicit way to express that. This way an Error has to be thrown, which in my opinion is  the correct behavior. If you pass the Option to the caller, you rely on correct handling.
